python/ray/ml/preprocessor.py

```diff
@@ -87,12 +93,22 @@ def transform_batch(self, df: DataBatchType) -> DataBatchType:
         Returns:
             DataBatchType: The transformed data batch.
         """
-        if self._is_fittable and not self.check_is_fitted():
+        if self.should_fit():
             raise PreprocessorNotFittedException(
                 "`fit` must be called before `transform_batch`."
             )
         return self._transform_batch(df)
 
+    def should_fit(self):
```

Comment:
Hmmm, got it. We can decide on the naming. But semantics are basically:
1. `fittable` is an inherent attribute of the "type" of the preprocessors. It also implies whether a `fit` method in meaningful at all throughout the entire lifetime of this preprocessor. 
2. should_fit/can_fit depends on the state a preprocessor is currently in (assuming it's fittable).

Exposing `check_is_fitted` alone is not enough, as you can see in `trainer.py` - it only checks for `check_is_fitted` in current impl, which leads to crash in the case of non-fittable preprocessors. That's why the proposal is to add `should_fit`.

`check_is_fitted` v.s. `is_fitted` or `can` v.s. `should` - I don't have much preference. @clarkzinzow @matthewdeng maybe as original author of the API?

Reply:
agree with the change. I am just nit-picking the naming.
hope to get things named consistently. thanks :)

Reply:
I opted for `should_fit` over `can_fit` since it's not indicating an optional operation for fittable preprocessors, it's a necessary operation: if a fittable preprocessor is not fit before calling `.transform()`, it will fail. An argument could be made for `need_fit`.

Reply:
WE can change `check_is_fitted` to `is_fitted`.

`should_fit` functionally is a bit strange to me, at least as a public API. In particular, I want to avoid the case where the user does something like:

```python
if (preprocessor.should_fit()):
    preprocessor.fit()
```

It's not clear how to differentiate the case where the preprocessor is fitted from the case where the preprocessor was already fitted before.

Reply:
@matthewdeng hmm, mind elaborating a bit more?
so 3 cases:
1. not fittable
2. fittable and fitted
3. fittable and not fitted yet

should_fit == case 3
if_fitted == 2 conditioned on (2 + 3)

Another way that I can see to work is just to enforce "at most once" fitting semantics internally - and caller doesn't have to call `should_fit` before `fit`. Which one do you prefer? Or are you proposing an alternative?

Reply:
I think I lack the context why you want to bundle these 2 things together in the first place.
but in my mind, the most intuitive way is to:

if calling fit() or fit_transform(), and not fittable, throw exception.
if calling fit(), and already fitted, print warning msg, and no-op.
if calling fit_transoform(), and already fitted, print warning msg, then proceed to transform().

Reply:
Can we raise an exception in the `fit()/fit_transfomr()` when already fitted instead? Logging is better than no logging, but I worry the behavior here isn't clear for users (I can see users thinking it should re-fit).